k2/csrc/ragged_ops_inl.h

```diff
@@ -566,24 +568,67 @@ void ArgMaxPerSublist(Ragged<T> &src, T initial_value, Array1<int32_t> *dst) {
     }
   } else {
     K2_CHECK_EQ(c->GetDeviceType(), kCuda);
-    argmax_internal::PairInputIterator<T> input_iter(values_data);
-    argmax_internal::PairOutputIterator<T> output_iter(output_data);
-    argmax_internal::PairMaxOp<T> op;
-    argmax_internal::Pair<T> initial_pair{initial_value, -1};
 
-    // This code is based on the example here:
-    // https://nvlabs.github.io/cub/structcub_1_1_device_segmented_reduce.html
+    size_t align = alignof(cub::KeyValuePair<int, T>);
+    Array1<int8_t> out_storage(c,
+        num_rows * sizeof(cub::KeyValuePair<int, T>) + align);
+    size_t shift = reinterpret_cast<std::intptr_t>(out_storage.Data()) % align;
+    shift = shift ? align - shift : 0;
+
+    cub::KeyValuePair<int, T> *d_out_data =
+      reinterpret_cast<cub::KeyValuePair<int, T>*>(out_storage.Data() + shift);
+
     std::size_t temp_storage_bytes = 0;
 
-    // the first time is to determine temporary device storage requirements
-    K2_CUDA_SAFE_CALL(cub::DeviceSegmentedReduce::Reduce(
-        nullptr, temp_storage_bytes, input_iter, output_iter, num_rows,
-        row_splits, row_splits + 1, op, initial_pair, c->GetCudaStream()));
+    K2_CUDA_SAFE_CALL(cub::DeviceSegmentedReduce::ArgMax(
+        nullptr, temp_storage_bytes, values_data, d_out_data, num_rows,
+        row_splits, row_splits + 1));
     Array1<int8_t> d_temp_storage(c, temp_storage_bytes);
-    K2_CUDA_SAFE_CALL(cub::DeviceSegmentedReduce::Reduce(
-        d_temp_storage.Data(), temp_storage_bytes, input_iter, output_iter,
-        num_rows, row_splits, row_splits + 1, op, initial_pair,
-        c->GetCudaStream()));
+    K2_CUDA_SAFE_CALL(cub::DeviceSegmentedReduce::ArgMax(
+        d_temp_storage.Data(), temp_storage_bytes, values_data, d_out_data,
+        num_rows, row_splits, row_splits + 1));
+
+    K2_EVAL(
+        c, num_rows, lambda_add_offset, (int32_t idx0) -> void {
+          if (row_splits[idx0 + 1] == row_splits[idx0])
+            output_data[idx0] = -1;
+          else
+            output_data[idx0] = d_out_data[idx0].key + row_splits[idx0];
+        });
+
+
+    // TODO: Finally, we will remove the checking below.
+    // To check whether the argmax result of cub is right. If there are errors
+    // of argmax results, we would recompute the error ones.
+
+    Array1<int8_t> flag(c, 1, 0);
+    int8_t* flag_data = flag.Data();
+    K2_EVAL(
+        c, src.NumElements(), lambda_check_argmax, (int32_t idx01) -> void {
+          int32_t idx0 = row_ids[idx01];
+          if (output_data[idx0] == -1) return;
+          T max_value = values_data[output_data[idx0]];
+          if (values_data[idx01] > max_value) {
+            flag_data[0] = 1;
```

Comment:
Shall we remove `flag` and print the warning inside the `if`?
We can use a guard, i.e.,`idx1 == 0`, to let each sublist print at most one warning.

---

`flag` involves one cpu->gpu, and one gpu->cpu call, no matter whether there is an error in `ArgMax` or not.

Reply:
It is hard to print the warning message inside the `if`, if we only allow to print once. We never know which element will go into the `if`, so `idx1==0` won't work.

I don't know the transferring cost between cpu and gpu precisely, I thought there was only one Byte. We will finally remove these code in one month or two, if we won't see any printout warnings.

Reply:
Then can we let the thread with "idx01" corresponding to the one having max value in a sub list print the warning?

Launching a kernel takes more time than that of transferring just one byte, I believe.

Reply:
You guys do it how you want.
My two cents, though:
 - If it's an error condition, printing multiple warnings is fine.
 - Transferring a byte from GPU->CPU is actually extremely slow, much slower than launching a kernel, because it forces the GPU to be synced.  A kernel launch just gets queued and control is returned to the CPU without anything necessarily being run at all (at least, not immediately).